File: preset_dump/src/services/program_data.py
"""Program data structure for Minilogue XD"""

from dataclasses import dataclass
from typing import Optional
from .enums import *


@dataclass
class ProgramData:
    """
    Adapted from the Minilogue XD MIDI Implementation Guide: 
    https://www.korg.com/us/support/download/manual/0/811/4440/
    Revision 1.01 (2020.2.10)
    """
    
    # Must be "PROG"
    header: Optional[str] = None
    
    # Up to 12 ASCII Characters for the Program Name
    program_name: Optional[str] = None
    
    # Octave: -2 to +2 (stored as 0-4)
    octave: int = 0
    
    portamento: int = 0
    
    # Key Trig On or Off (added in Firmware 2.0)
    key_trig: bool = False
    
    voice_mode_depth: int = 0
    voice_mode_type: VoiceModeType = VoiceModeType.NONE
    
    # VCO 1
    vco1_wave: VcoWave = VcoWave.SQR
    vco1_octave: int = 0  # 0~3=16',8',4',2'
    vco1_pitch: int = 0
    vco1_shape: int = 0
    
    # VCO 2
    vco2_wave: VcoWave = VcoWave.SQR
    vco2_octave: int = 0  # 0~3=16',8',4',2'
    vco2_pitch: int = 0
    vco2_shape: int = 0
    
    oscillator_sync: bool = False
    ring_mod: bool = False
    cross_mod_depth: int = 0
    
    # Multi Engine
    multi_osc_type: MultiOscType = MultiOscType.NOISE
    selected_multi_osc_noise: MultiOscNoise = MultiOscNoise.HIGH
    selected_multi_osc_vpm: MultiOscVPM = MultiOscVPM.SIN1
    selected_multi_osc_user: int = 0  # 0-15
    
    shape_noise: int = 0
    shape_vpm: int = 0
    shape_user: int = 0
    shift_shape_noise: int = 0
    shift_shape_vpm: int = 0
    shift_shape_user: int = 0
    
    # Mixer
    vco1_level: int = 0
    vco2_level: int = 0
    multi_level: int = 0
    
    # Filter
    filter_cutoff: int = 0
    filter_resonance: int = 0
    filter_cutoff_drive: int = 0  # 0-2
    filter_cutoff_keyboard_track: int = 0  # 0-2
    
    # Amp EG
    amp_eg_attack: int = 0
    amp_eg_decay: int = 0
    amp_eg_sustain: int = 0
    amp_eg_release: int = 0
    
    # EG
    eg_attack: int = 0
    eg_decay: int = 0
    eg_int: int = 0
    eg_target: EGTarget = EGTarget.CUTOFF
    
    # LFO
    lfo_wave: VcoWave = VcoWave.SQR
    lfo_mode: LFOMode = LFOMode.NORMAL
    lfo_rate: int = 0
    lfo_int: int = 0
    lfo_target: LFOTarget = LFOTarget.CUTOFF
    
    # Mod FX
    mod_fx_on_off: bool = False
    mod_fx_type: ModFxType = ModFxType.NONE
    mod_fx_chorus: ModFxChorus = ModFxChorus.STEREO
    mod_fx_ensemble: ModFxEnsemble = ModFxEnsemble.STEREO
    mod_fx_phaser: ModFxPhaser = ModFxPhaser.STEREO
    mod_fx_flanger: ModFxFlanger = ModFxFlanger.STEREO
    mod_fx_user: int = 0  # 0-15
    mod_fx_time: int = 0
    mod_fx_depth: int = 0
    
    # Delay FX
    delay_on_off: bool = False
    delay_sub_type: DelaySubType = DelaySubType.STEREO
    delay_time: int = 0
    delay_depth: int = 0
    
    # Reverb FX
    reverb_on_off: bool = False
    reverb_sub_type: ReverbSubType = ReverbSubType.HALL
    reverb_time: int = 0
    reverb_depth: int = 0
    
    # Bend & Joystick
    bend_range_plus: int = 0  # OFF~+12Note
    bend_range_minus: int = 0  # OFF~-12Note
    joystick_assign_plus: AssignTarget = AssignTarget.GATE_TIME
    joystick_range_plus: int = 0  # 0~200=-100%~+100%
    joystick_assign_minus: AssignTarget = AssignTarget.GATE_TIME
    joystick_range_minus: int = 0  # 0~200=-100%~+100%
    
    # CV In
    cv_in_mode: CVInMode = CVInMode.MODULATION
    cv_in1_assign: AssignTarget = AssignTarget.GATE_TIME
    cv_in1_range: int = 0  # 0~200=-100%~+100%
    cv_in2_assign: AssignTarget = AssignTarget.GATE_TIME
    cv_in2_range: int = 0  # 0~200=-100%~+100%
    
    # Tuning
    micro_tuning: MicroTuning = MicroTuning.EQUAL_TEMP
    scale_key: int = 0  # 0~24=-12Note~+12Note
    program_tuning: int = 0  # 0~100=-50Cent~+50Cent
    
    # LFO Advanced
    lfo_key_sync: bool = False
    lfo_voice_sync: bool = False
    lfo_target_osc: LFOTargetOsc = LFOTargetOsc.ALL
    
    # Modulation
    cutoff_velocity: int = 0
    amp_velocity: int = 0
    
    # Multi Engine Advanced
    multi_octave: int = 0  # 0~3=16',8',4',2'
    multi_routing: MultiRouting = MultiRouting.PRE_VCF
    
    # EG Advanced
    eg_legato: bool = False
    
    # Portamento Advanced
    portamento_mode: PortamentoMode = PortamentoMode.AUTO
    portamento_bpm_sync: bool = False
    
    # Program Level
    program_level: int = 72  # 12~132=-18dB~+6dB
    
    # VPM Parameters (0~200=-100%~+100%)
    vpm_parameter1_feedback: int = 100
    vpm_parameter2_noise_depth: int = 100
    vpm_parameter3_shape_mod_int: int = 100
    vpm_parameter4_mod_attack: int = 100
    vpm_parameter5_mod_decay: int = 100
    vpm_parameter6_mod_key_track: int = 100
    
    # User Parameters
    user_param1: int = 0
    user_param2: int = 0
    user_param3: int = 0
    user_param4: int = 0
    user_param5: int = 0
    user_param6: int = 0
    user_param56_type: int = 0
    user_param1234_type: int = 0
    
    user_param1_type: UserParamType = UserParamType.PERCENT_TYPE
    user_param2_type: UserParamType = UserParamType.PERCENT_TYPE
    user_param3_type: UserParamType = UserParamType.PERCENT_TYPE
    user_param4_type: UserParamType = UserParamType.PERCENT_TYPE
    user_param5_type: UserParamType = UserParamType.PERCENT_TYPE
    user_param6_type: UserParamType = UserParamType.PERCENT_TYPE
    
    # Program Transpose
    program_transpose: int = 13  # -12~+12 Note (stored as 1-25)
    
    # Dry/Wet
    delay_dry_wet: int = 0
    reverb_dry_wet: int = 0
    
    # MIDI After Touch
    midi_after_touch_assign: AssignTarget = AssignTarget.GATE_TIME
    
    # Must be "PRED"
    program_end_marker: Optional[str] = None

File: ts_web/src/program-data.ts
import * as Enums from './enums';

export class ProgramData {
  header?: string;
  programName?: string;
  octave: number = 0;
  portamento: number = 0;
  keyTrig: boolean = false;
  voiceModeDepth: number = 0;
  voiceModeType: Enums.VoiceModeType = Enums.VoiceModeType.NONE;

  // VCO 1
  vco1Wave: Enums.VcoWave = Enums.VcoWave.SQR;
  vco1Octave: number = 0;
  vco1Pitch: number = 0;
  vco1Shape: number = 0;

  // VCO 2
  vco2Wave: Enums.VcoWave = Enums.VcoWave.SQR;
  vco2Octave: number = 0;
  vco2Pitch: number = 0;
  vco2Shape: number = 0;

  oscillatorSync: boolean = false;
  ringMod: boolean = false;
  crossModDepth: number = 0;

  // Multi Engine
  multiOscType: Enums.MultiOscType = Enums.MultiOscType.NOISE;
  selectedMultiOscNoise: Enums.MultiOscNoise = Enums.MultiOscNoise.HIGH;
  selectedMultiOscVpm: Enums.MultiOscVPM = Enums.MultiOscVPM.SIN1;
  selectedMultiOscUser: number = 0;

  shapeNoise: number = 0;
  shapeVpm: number = 0;
  shapeUser: number = 0;
  shiftShapeNoise: number = 0;
  shiftShapeVpm: number = 0;
  shiftShapeUser: number = 0;

  // Mixer
  vco1Level: number = 0;
  vco2Level: number = 0;
  multiLevel: number = 0;

  // Filter
  filterCutoff: number = 0;
  filterResonance: number = 0;
  filterCutoffDrive: number = 0;
  filterCutoffKeyboardTrack: number = 0;

  // Amp EG
  ampEgAttack: number = 0;
  ampEgDecay: number = 0;
  ampEgSustain: number = 0;
  ampEgRelease: number = 0;

  // EG
  egAttack: number = 0;
  egDecay: number = 0;
  egInt: number = 0;
  egTarget: Enums.EGTarget = Enums.EGTarget.CUTOFF;

  // LFO
  lfoWave: Enums.VcoWave = Enums.VcoWave.SQR;
  lfoMode: Enums.LFOMode = Enums.LFOMode.NORMAL;
  lfoRate: number = 0;
  lfoInt: number = 0;
  lfoTarget: Enums.LFOTarget = Enums.LFOTarget.CUTOFF;

  // Mod FX
  modFxOnOff: boolean = false;
  modFxType: Enums.ModFxType = Enums.ModFxType.NONE;
  modFxChorus: Enums.ModFxChorus = Enums.ModFxChorus.STEREO;
  modFxEnsemble: Enums.ModFxEnsemble = Enums.ModFxEnsemble.STEREO;
  modFxPhaser: Enums.ModFxPhaser = Enums.ModFxPhaser.STEREO;
  modFxFlanger: Enums.ModFxFlanger = Enums.ModFxFlanger.STEREO;
  modFxUser: number = 0;
  modFxTime: number = 0;
  modFxDepth: number = 0;

  // Delay FX
  delayOnOff: boolean = false;
  delaySubType: Enums.DelaySubType = Enums.DelaySubType.STEREO;
  delayTime: number = 0;
  delayDepth: number = 0;

  // Reverb FX
  reverbOnOff: boolean = false;
  reverbSubType: Enums.ReverbSubType = Enums.ReverbSubType.HALL;
  reverbTime: number = 0;
  reverbDepth: number = 0;

  // Bend & Joystick
  bendRangePlus: number = 0;
  bendRangeMinus: number = 0;
  joystickAssignPlus: Enums.AssignTarget = Enums.AssignTarget.GATE_TIME;
  joystickRangePlus: number = 0;
  joystickAssignMinus: Enums.AssignTarget = Enums.AssignTarget.GATE_TIME;
  joystickRangeMinus: number = 0;

  // CV In
  cvInMode: Enums.CVInMode = Enums.CVInMode.MODULATION;
  cvIn1Assign: Enums.AssignTarget = Enums.AssignTarget.GATE_TIME;
  cvIn1Range: number = 0;
  cvIn2Assign: Enums.AssignTarget = Enums.AssignTarget.GATE_TIME;
  cvIn2Range: number = 0;

  // Tuning
  microTuning: Enums.MicroTuning = Enums.MicroTuning.EQUAL_TEMP;
  scaleKey: number = 0;
  programTuning: number = 0;

  // LFO Advanced
  lfoKeySync: boolean = false;
  lfoVoiceSync: boolean = false;
  lfoTargetOsc: Enums.LFOTargetOsc = Enums.LFOTargetOsc.ALL;

  // Modulation
  cutoffVelocity: number = 0;
  ampVelocity: number = 0;

  // Multi Engine Advanced
  multiOctave: number = 0;
  multiRouting: Enums.MultiRouting = Enums.MultiRouting.PRE_VCF;

  // EG Advanced
  egLegato: boolean = false;

  // Portamento Advanced
  portamentoMode: Enums.PortamentoMode = Enums.PortamentoMode.AUTO;
  portamentoBpmSync: boolean = false;

  // Program Level
  programLevel: number = 72;

  // VPM Parameters
  vpmParameter1Feedback: number = 100;
  vpmParameter2NoiseDepth: number = 100;
  vpmParameter3ShapeModInt: number = 100;
  vpmParameter4ModAttack: number = 100;
  vpmParameter5ModDecay: number = 100;
  vpmParameter6ModKeyTrack: number = 100;

  // User Parameters
  userParam1: number = 0;
  userParam2: number = 0;
  userParam3: number = 0;
  userParam4: number = 0;
  userParam5: number = 0;
  userParam6: number = 0;
  userParam56Type: number = 0;
  userParam1234Type: number = 0;

  userParam1Type: Enums.UserParamType = Enums.UserParamType.PERCENT_TYPE;
  userParam2Type: Enums.UserParamType = Enums.UserParamType.PERCENT_TYPE;
  userParam3Type: Enums.UserParamType = Enums.UserParamType.PERCENT_TYPE;
  userParam4Type: Enums.UserParamType = Enums.UserParamType.PERCENT_TYPE;
  userParam5Type: Enums.UserParamType = Enums.UserParamType.PERCENT_TYPE;
  userParam6Type: Enums.UserParamType = Enums.UserParamType.PERCENT_TYPE;

  // Program Transpose
  programTranspose: number = 13;

  // Dry/Wet
  delayDryWet: number = 0;
  reverbDryWet: number = 0;

  // MIDI After Touch
  midiAfterTouchAssign: Enums.AssignTarget = Enums.AssignTarget.GATE_TIME;

  programEndMarker?: string;
}